chain/bitcoind.go

```diff
@@ -0,0 +1,1122 @@
+		rescanUpdate:        make(chan interface{}),
+		watchOutPoints:      make(map[wire.OutPoint]struct{}),
+		watchAddrs:          make(map[string]struct{}),
+		watchTxIDs:          make(map[chainhash.Hash]struct{}),
+		quit:                make(chan struct{}),
+		memPool:             make(map[chainhash.Hash]struct{}),
+		memPoolExp:          make(map[int32]map[chainhash.Hash]struct{}),
+	}
+	rpcClient, err := rpcclient.New(client.connConfig, nil)
+	if err != nil {
+		return nil, err
+	}
+	client.client = rpcClient
+	return client, nil
+}
+
+// BackEnd returns the name of the driver.
+func (c *BitcoindClient) BackEnd() string {
+	return "bitcoind"
+}
+
+// GetCurrentNet returns the network on which the bitcoind instance is running.
+func (c *BitcoindClient) GetCurrentNet() (wire.BitcoinNet, error) {
+	hash, err := c.client.GetBlockHash(0)
+	if err != nil {
+		return 0, err
+	}
+
+	switch *hash {
+	case *chaincfg.TestNet3Params.GenesisHash:
+		return chaincfg.TestNet3Params.Net, nil
+	case *chaincfg.RegressionNetParams.GenesisHash:
+		return chaincfg.RegressionNetParams.Net, nil
+	case *chaincfg.MainNetParams.GenesisHash:
+		return chaincfg.MainNetParams.Net, nil
+	default:
+		return 0, errors.New("unknown network")
+	}
+}
+
+// GetBestBlock returns the highest block known to bitcoind.
+func (c *BitcoindClient) GetBestBlock() (*chainhash.Hash, int32, error) {
+	bcinfo, err := c.client.GetBlockChainInfo()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	hash, err := chainhash.NewHashFromStr(bcinfo.BestBlockHash)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return hash, bcinfo.Blocks, nil
+}
+
+// GetBlockHeight returns the height for the hash, if known, or returns an
+// error.
+func (c *BitcoindClient) GetBlockHeight(hash *chainhash.Hash) (int32, error) {
+	header, err := c.GetBlockHeaderVerbose(hash)
+	if err != nil {
+		return 0, err
+	}
+
+	return header.Height, nil
+}
+
+// GetBlock returns a block from the hash.
+func (c *BitcoindClient) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock,
+	error) {
+	return c.client.GetBlock(hash)
+}
+
+// GetBlockVerbose returns a verbose block from the hash.
+func (c *BitcoindClient) GetBlockVerbose(hash *chainhash.Hash) (
+	*btcjson.GetBlockVerboseResult, error) {
+	return c.client.GetBlockVerbose(hash)
+}
+
+// GetBlockHash returns a block hash from the height.
+func (c *BitcoindClient) GetBlockHash(height int64) (*chainhash.Hash, error) {
+	return c.client.GetBlockHash(height)
+}
+
+// GetBlockHeader returns a block header from the hash.
+func (c *BitcoindClient) GetBlockHeader(
+	hash *chainhash.Hash) (*wire.BlockHeader, error) {
+	return c.client.GetBlockHeader(hash)
+}
+
+// GetBlockHeaderVerbose returns a block header from the hash.
+func (c *BitcoindClient) GetBlockHeaderVerbose(hash *chainhash.Hash) (
+	*btcjson.GetBlockHeaderVerboseResult, error) {
+	return c.client.GetBlockHeaderVerbose(hash)
+}
+
+// GetRawTransactionVerbose returns a transaction from the tx hash.
```

Comment:
Do all these actually need to be exported? I thought the point of this struct (and the greater interface) was just to provide an abstraction that `btcwallet` can use to interact with the chain in a generic manner. 

At this rate, why not just _embed_ the underlying rpc client? IIRC, the client for the btcd backend does the same.

Reply:
Many of the APIs implemented by `btcd` are no longer available in `bitcoind`, so I've had to emulate them, in a similar way to what I did with `neutrino`. The greater interface's methods are what `btcwallet` uses from `btcd`, and both `bitcoind` and `neutrino` emulate those to avoid making larger changes to `btcwallet`. We can redesign the chain interface and update its dependencies in additional future housecleaning.

Reply:
👍